allow model types to be ignored during object deletion

mtbc opened the pull request:
# What this PR does

This PR adds a `Delete2.typesToIgnore` property that can be used to stall graph traversal at nodes of those types. This is a strictly *experimental* feature that when misused has the potential to leave the model graph in an unfortunate state or to produce strange error messages. It is however a useful interim feature for IDR workflows until we upgrade Hibernate and adjust its query caching.

# Testing this PR

Set graphs logging to DEBUG using https://docs.openmicroscopy.org/omero/5.4.0-m3/developers/Server/ObjectGraphs.html#changing-the-log-level. (Alternatively, @manics has means to instrument PostgreSQL queries.)

Link a map annotation to multiple images. Using a Python script delete one of the links via `Delete2(..., typesToIgnore=['Annotation'])`. Observe that with the extra `typesToIgnore` a lot less processing than otherwise happens with the annotation's *other* links.

Also see https://github.com/openmicroscopy/openmicroscopy/pull/5460#issuecomment-324338672.

Integration tests at https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/lastCompletedBuild/testngreports/integration.delete/AnnotationDeleteTest/.

# Related reading

https://trello.com/c/mfW0N1an/89-poor-performance-when-deleting-many-one-annotation-links

mtbc committed:
add new Delete2.typesToIgnore option in object deletion

mtbc committed:
allow ignoring of certain model classes when deleting objects

manics commented:
Functionally `ChildOption(excludeType=['Annotation'])` and `typesToIgnore=['Annotation']` have the same effect. At the postgres level the queries seem to be similar- in both cases when deleting `DatasetAnnotationLink`s other `xxxannotationlink` are still queried. After enabling DEBUG for graphs:
```
  <logger name="omero.cmd.graphs" level="DEBUG"/>
  <logger name="ome.services.graphs" level="DEBUG"/>
```
and extracting the logs from Blitz whilst deleting annotations with/without `excludeType` and `typesToIgnore`: `docker-compose exec omero tail -n0 -F /opt/omero/server/OMERO.server/var/log/Blitz-0.log | grep graph --line-buffered`

This is a summary of counts of each log message (extracted using https://gist.github.com/manics/9140738):
```
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched A,L so will review ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT for rule L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a} ({'graphs-1010-00.txt': 20})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched IAnnotationLink[D].child = A:MapAnnotation[E]{i}/d to A:{r}, where A is ome.model.annotations.MapAnnotation[[h]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a}, where A is ome.model.annotations.MapAnnotation[[h]]/RELEVANT, L is ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010-00.txt': 20})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010-00.txt': 20, 'graphs-1010-10.txt': 20, 'graphs-1010-11.txt': 20})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010-00.txt': 20})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) marked ome.model.annotations.MapAnnotation[[h]] as IS_LAST ({'graphs-1010-01.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) marked ome.model.annotations.MapAnnotation[[h]] as RELEVANT to verify IS_NOT_LAST status ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].child links to ome.model.annotations.MapAnnotation[[h]] ({'graphs-1010-00.txt': 200, 'graphs-1010-10.txt': 200, 'graphs-1010-01.txt': 200, 'graphs-1010-11.txt': 200})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].parent links to ome.model.containers.Dataset[[h]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].child links to ome.model.annotations.MapAnnotation[[h]] ({'graphs-1010-00.txt': 20, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].parent links to ome.model.containers.Project[[i]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Dataset[[h]].annotationLinks links to ome.model.annotations.DatasetAnnotationLink[[h]] ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Project[[i]].annotationLinks links to ome.model.annotations.ProjectAnnotationLink[[h]] ({'graphs-1010-00.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) processor: delete ome.model.annotations.DatasetAnnotationLink[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]] ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/DELETE ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/DELETE for error conditions ({'graphs-1010-00.txt': 100, 'graphs-1010-10.txt': 100, 'graphs-1010-01.txt': 100, 'graphs-1010-11.txt': 100})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_LAST ({'graphs-1010-01.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_LAST for error conditions ({'graphs-1010-01.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010-00.txt': 20, 'graphs-1010-10.txt': 20, 'graphs-1010-11.txt': 20})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST for error conditions ({'graphs-1010-00.txt': 10, 'graphs-1010-10.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010-00.txt': 20, 'graphs-1010-10.txt': 10, 'graphs-1010-01.txt': 10, 'graphs-1010-11.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010-00.txt': 22})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT for error conditions ({'graphs-1010-00.txt': 10})
DEBUG [ omero.cmd.graphs.ChildOptionsPolicy] ([i]-thread-[i]) excluding all children of its type, so making ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010-10.txt': 20, 'graphs-1010-11.txt': 20})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) Request callContext: {omero.group=-[i]} ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) User callContext: {omero.client.uuid=[session], omero.session.uuid=[session]} ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]]}, childOptions = [<excludeType = [Annotation]>], dryRun = false ({'graphs-1010-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]]}, childOptions = [<excludeType = [Annotation]>], typesToIgnore = [Annotation], dryRun = false ({'graphs-1010-11.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]]}, dryRun = false ({'graphs-1010-00.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]]}, typesToIgnore = [Annotation], dryRun = false ({'graphs-1010-01.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) response: deletedObjects = {ome.model.annotations.DatasetAnnotationLink:[[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h],[h]]} ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/IS_LAST ({'graphs-1010-01.txt': 20})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010-11.txt': 40})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010-01.txt': 10})
INFO [ ome.services.util.ServiceHandler] ([i]-thread-[i]) Executor.doWork -- omero.cmd.HandleI.run[[session]/IHandle[session], omero.cmd.graphs.Delete2I@[h]] ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (.Server-[i]) Add callback: ;>*h2-sC^eP2\/(Uy^zlL/[session] ({'graphs-1010-01.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (.Server-[i]) Closing... ({'graphs-1010-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (.Server-[i]) getResponse: null ({'graphs-1010-00.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (.Server-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs-1010-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) in delete of [i], deleted [i] in total ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs-1010-00.txt': 1, 'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify step [i] of [i] ({'graphs-1010-00.txt': 6, 'graphs-1010-10.txt': 6, 'graphs-1010-01.txt': 6, 'graphs-1010-11.txt': 6})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: Y$?=biZXMiv.d?A@+\'jX/[session] ({'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: bS1aLV&f(B0M|JIn<*K3/[session] ({'graphs-1010-00.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: uK>[i]R%[i][+#F*H3J8H+$_/[session] ({'graphs-1010-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Closing... ({'graphs-1010-00.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) getResponse: null ({'graphs-1010-10.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs-1010-00.txt': 1, 'graphs-1010-01.txt': 1, 'graphs-1010-11.txt': 1})
```
Test script: https://gist.github.com/manics/8ebc8f4dac34d1ba313f3032d6ec5c41#file-
- `graphs-1010-00.txt`: No delete options (`./test_delete_performance.py 10 10 0 0`)
- `graphs-1010-01.txt`: excludeType (`./test_delete_performance.py 10 10 1 0`)
- `graphs-1010-10.txt`: typesToIgnore (`./test_delete_performance.py 10 10 0 1`)
- `graphs-1010-11.txt`: excludeType and typesToIgnore (`./test_delete_performance.py 10 10 1 1`)

manics commented:
I'll attempt to test the actual performance of this PR on a copy of the IDR since it's difficult to tell from the logs what's changed.

mtbc commented:
Maybe `graphs-1010-01.txt` is the `typesToIgnore`.

mtbc commented:
Might be interesting to try deleting just a few of the `DatasetAnnotationLink` at a time, make sure there's no "reviewing" line for the untargeted ones.

manics commented:
--rebased-to #5464

manics commented:
> Maybe graphs-1010-01.txt is the typesToIgnore.

You're correct, I've fixed the comment

manics commented:
Results, this time deleting just one `DatasetAnnotationLink` (`linkids = linkids[:1]`)
```
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched A,L so will review ome.model.annotations.DatasetAnnotationLink[[h]]/IRRELEVANT for rule L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a} ({'graphs-1010d1-00.txt': 18})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched A,L so will review ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT for rule L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a} ({'graphs-1010d1-00.txt': 2})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched IAnnotationLink[D].child = A:MapAnnotation[E]{i}/d to A:{r}, where A is ome.model.annotations.MapAnnotation[[h]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a}, where A is ome.model.annotations.MapAnnotation[[h]]/RELEVANT, L is ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010d1-00.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.DatasetAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010d1-00.txt': 18})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010d1-11.txt': 2, 'graphs-1010d1-00.txt': 2, 'graphs-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010d1-00.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) marked ome.model.annotations.MapAnnotation[[h]] as IS_LAST ({'graphs-1010d1-01.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) marked ome.model.annotations.MapAnnotation[[h]] as RELEVANT to verify IS_NOT_LAST status ({'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].child links to ome.model.annotations.MapAnnotation[[h]] ({'graphs-1010d1-01.txt': 11, 'graphs-1010d1-11.txt': 11, 'graphs-1010d1-00.txt': 20, 'graphs-1010d1-10.txt': 11})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[h]].parent links to ome.model.containers.Dataset[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].child links to ome.model.annotations.MapAnnotation[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 2, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[h]].parent links to ome.model.containers.Project[[i]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Dataset[[h]].annotationLinks links to ome.model.annotations.DatasetAnnotationLink[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 10, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Project[[i]].annotationLinks links to ome.model.annotations.ProjectAnnotationLink[[h]] ({'graphs-1010d1-00.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) processor: delete ome.model.annotations.DatasetAnnotationLink[[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/DELETE ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/DELETE for error conditions ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010d1-00.txt': 18})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[h]]/IRRELEVANT for error conditions ({'graphs-1010d1-00.txt': 9})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_LAST ({'graphs-1010d1-01.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_LAST for error conditions ({'graphs-1010d1-01.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010d1-11.txt': 2, 'graphs-1010d1-00.txt': 2, 'graphs-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST for error conditions ({'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 2, 'graphs-1010d1-00.txt': 2, 'graphs-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT ({'graphs-1010d1-00.txt': 4})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[h]]/IRRELEVANT for error conditions ({'graphs-1010d1-00.txt': 1})
DEBUG [ omero.cmd.graphs.ChildOptionsPolicy] ([i]-thread-[i]) excluding all children of its type, so making ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010d1-11.txt': 2, 'graphs-1010d1-10.txt': 2})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) Request callContext: {omero.group=-[i]} ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) User callContext: {omero.client.uuid=[session], omero.session.uuid=[session]} ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h]]}, childOptions = [<excludeType = [Annotation]>], dryRun = false ({'graphs-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h]]}, childOptions = [<excludeType = [Annotation]>], typesToIgnore = [Annotation], dryRun = false ({'graphs-1010d1-11.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h]]}, dryRun = false ({'graphs-1010d1-00.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[h]]}, typesToIgnore = [Annotation], dryRun = false ({'graphs-1010d1-01.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) response: deletedObjects = {ome.model.annotations.DatasetAnnotationLink:[[h]]} ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/IS_LAST ({'graphs-1010d1-01.txt': 2})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/IS_NOT_LAST ({'graphs-1010d1-11.txt': 5})
DEBUG [ omero.cmd.graphs.SkipTailPolicy] ([i]-thread-[i]) halting review at ome.model.annotations.MapAnnotation[[h]]/RELEVANT ({'graphs-1010d1-01.txt': 1})
INFO [ ome.services.util.ServiceHandler] ([i]-thread-[i]) Executor.doWork -- omero.cmd.HandleI.run[[session]/IHandle[session], omero.cmd.graphs.Delete2I@[h]] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) Add callback: Bu7[&C7Q4835A3@UW,~[i]/[session] ({'graphs-1010d1-01.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) Add callback: ]S^lg.:EF|[i])&hC.PU{h/[session] ({'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) Add callback: oIm\'\"]WRU03M(S}\'[i]%^&/[session] ({'graphs-1010d1-00.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) Add callback: yx)k^nO*[i]yOZ=p(]M^;(/[session] ({'graphs-1010d1-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) Closing... ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) getResponse: null ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (Server-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) in delete of [i], deleted [i] in total ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs-1010d1-01.txt': 1, 'graphs-1010d1-11.txt': 1, 'graphs-1010d1-00.txt': 1, 'graphs-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify step [i] of [i] ({'graphs-1010d1-01.txt': 6, 'graphs-1010d1-11.txt': 6, 'graphs-1010d1-00.txt': 6, 'graphs-1010d1-10.txt': 6})
```

mtbc commented:
That may not be as good as I had hoped: could you make at least the `ome.services.graphs.GraphTraversal` bit of the ordered/unsummarized log available? This afternoon I shall take a look to see if there is still more processing occurring than I would have liked in the `graphs-1010d1-01.txt` case.

mtbc committed:
fix Delete2 constructor in MATLAB code

new field typesToIgnore

raises question of needing an
	omero.gateway.util.Requests.delete()
for MATLAB folks too

manics commented:
[graphs-1010d1-00.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245159/graphs-1010d1-00.txt)
[graphs-1010d1-01.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245161/graphs-1010d1-01.txt)
[graphs-1010d1-10.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245162/graphs-1010d1-10.txt)
[graphs-1010d1-11.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245160/graphs-1010d1-11.txt)

mtbc commented:
*Do* preserve how you test this but it looks as if a last-minute rethink I had unexpectedly reduced this PR's efficacy so feel free to hold off for now @manics and I shall push an improvement.

manics commented:
I'll write up full instructions as a gist and link them.

mtbc committed:
promptly push ignored objects outside

[E] => [O]

mtbc commented:
Fix pushed: one should now see a marked improvement from `typesToIgnore`.

manics commented:
> I'll write up full instructions as a gist and link them.

https://gist.github.com/manics/9bb031d4b8215bb1c6a09808ff73bf1b

mtbc commented:
The `test_delete_performance.py` script is fun, thank you. I adopted the `linkids = linkids[:1]` and am indeed seeing a faster delete with `typesToIgnore`.

manics commented:
I've tested the latest commits on the [`metadata53` branch](https://github.com/openmicroscopy/openmicroscopy/pull/5464) and it's looking more promising:
```
$ wc graphs2-1010d1-??.txt
     356    3867   68620 graphs2-1010d1-00.txt
      31     322    4924 graphs2-1010d1-01.txt
     296    3214   56796 graphs2-1010d1-10.txt
      31     327    4972 graphs2-1010d1-11.txt
```
```
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched IAnnotationLink[D].child = A:MapAnnotation[E]{i}/d to A:{r}, where A is ome.model.annotations.MapAnnotation[[i]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[i]]/RELEVANT ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[!O] to L:[-], where L is ome.model.annotations.DatasetAnnotationLink[[i]]/DELETE, making ome.model.annotations.DatasetAnnotationLink[[i]]/DELETE ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 9, 'graphs2-1010d1-10.txt': 9})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[!O] to L:[-], where L is ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT, making ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 72, 'graphs2-1010d1-10.txt': 54})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[!O] to L:[-], where L is ome.model.annotations.ProjectAnnotationLink[[i]]/IRRELEVANT, making ome.model.annotations.ProjectAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 8, 'graphs2-1010d1-10.txt': 6})
DEBUG [ ome.services.graphs.GraphPolicyRule] ([i]-thread-[i]) matched L:IAnnotationLink[E].parent = [E]{ia}, L.child = A:Annotation[E]{r} to A:{a}, where A is ome.model.annotations.MapAnnotation[[i]]/RELEVANT, L is ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT, making ome.model.annotations.MapAnnotation[[i]]/IS_NOT_LAST ({'graphs2-1010d1-00.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.DatasetAnnotationLink[[i]]/DELETE ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 9, 'graphs2-1010d1-10.txt': 9})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 72, 'graphs2-1010d1-10.txt': 54})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[i]]/IS_NOT_LAST ({'graphs2-1010d1-00.txt': 2, 'graphs2-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[i]]/OUTSIDE ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.MapAnnotation[[i]]/RELEVANT ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) adjusted ome.model.annotations.ProjectAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 8, 'graphs2-1010d1-10.txt': 6})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) marked ome.model.annotations.MapAnnotation[[i]] as RELEVANT to verify IS_NOT_LAST status ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[i]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[i]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[i]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.Annotation[[i]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].child links to ome.model.annotations.MapAnnotation[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 20, 'graphs2-1010d1-10.txt': 20})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.DatasetAnnotationLink[[i]].parent links to ome.model.containers.Dataset[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].child links to ome.model.annotations.MapAnnotation[[i]] ({'graphs2-1010d1-00.txt': 2, 'graphs2-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].details.creationEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].details.group links to ome.model.meta.ExperimenterGroup[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].details.owner links to ome.model.meta.Experimenter[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].details.updateEvent links to ome.model.meta.Event[[h]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.annotations.ProjectAnnotationLink[[i]].parent links to ome.model.containers.Project[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Dataset[[i]].annotationLinks links to ome.model.annotations.DatasetAnnotationLink[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 10, 'graphs2-1010d1-10.txt': 10})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) ome.model.containers.Project[[i]].annotationLinks links to ome.model.annotations.ProjectAnnotationLink[[i]] ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) processor: delete ome.model.annotations.DatasetAnnotationLink[[i]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[i]]/DELETE ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 5, 'graphs2-1010d1-10.txt': 5})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[i]]/DELETE for error conditions ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 36, 'graphs2-1010d1-10.txt': 18})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.DatasetAnnotationLink[[i]]/IRRELEVANT for error conditions ({'graphs2-1010d1-00.txt': 9, 'graphs2-1010d1-10.txt': 9})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[i]]/IS_NOT_LAST ({'graphs2-1010d1-00.txt': 2, 'graphs2-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[i]]/IS_NOT_LAST for error conditions ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.MapAnnotation[[i]]/RELEVANT ({'graphs2-1010d1-00.txt': 2, 'graphs2-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[i]]/IRRELEVANT ({'graphs2-1010d1-00.txt': 4, 'graphs2-1010d1-10.txt': 2})
DEBUG [ ome.services.graphs.GraphTraversal] ([i]-thread-[i]) reviewing ome.model.annotations.ProjectAnnotationLink[[i]]/IRRELEVANT for error conditions ({'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.ChildOptionsPolicy] ([i]-thread-[i]) excluding all children of its type, so making ome.model.annotations.MapAnnotation[[i]]/IS_NOT_LAST ({'graphs2-1010d1-10.txt': 2})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) Request callContext: {omero.group=-[i]} ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) User callContext: {omero.client.uuid=[session], omero.session.uuid=[session]} ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[i]]}, childOptions = [<excludeType = [Annotation]>], dryRun = false ({'graphs2-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[i]]}, childOptions = [<excludeType = [Annotation]>], typesToIgnore = [Annotation], dryRun = false ({'graphs2-1010d1-11.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[i]]}, dryRun = false ({'graphs2-1010d1-00.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) request: targetObjects = {DatasetAnnotationLink:[[i]]}, typesToIgnore = [Annotation], dryRun = false ({'graphs2-1010d1-01.txt': 1})
DEBUG [ omero.cmd.graphs.Delete2I] ([i]-thread-[i]) response: deletedObjects = {ome.model.annotations.DatasetAnnotationLink:[[i]]} ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
DEBUG [ omero.cmd.graphs.IgnoreTypePolicy] ([i]-thread-[i]) ignoring all objects of its type, so making ome.model.annotations.MapAnnotation[[i]]/OUTSIDE ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1})
INFO [ ome.services.util.ServiceHandler] ([i]-thread-[i]) Executor.doWork -- omero.cmd.HandleI.run[[session]/IHandle[session], omero.cmd.graphs.Delete2I@[h]] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) in delete of [i], deleted [i] in total ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] ([i]-thread-[i]) notify step [i] of [i] ({'graphs2-1010d1-01.txt': 6, 'graphs2-1010d1-11.txt': 6, 'graphs2-1010d1-00.txt': 6, 'graphs2-1010d1-10.txt': 6})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: #Un]![i]s.OC5+][i]@$R;|`/[session] ({'graphs2-1010d1-01.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: )FFo)iCO3\\^:^ZDt)Q3,/[session] ({'graphs2-1010d1-00.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: XJ;-TA\"\\_;:*MI8dKc]?/[session] ({'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Add callback: YA\/Yf`UR5*-w$m9d(~[a/[session] ({'graphs2-1010d1-11.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) Closing... ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) getResponse: null ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
INFO [ omero.cmd.graphs.Delete2I.@[h]] (l.Server-[i]) notify finished: omero.cmd.Delete2Response@[h]/omero.cmd.Status@[h] ({'graphs2-1010d1-01.txt': 1, 'graphs2-1010d1-11.txt': 1, 'graphs2-1010d1-00.txt': 1, 'graphs2-1010d1-10.txt': 1})
```
[graphs2-1010d1-00.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245842/graphs2-1010d1-00.txt)
[graphs2-1010d1-01.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245843/graphs2-1010d1-01.txt)
[graphs2-1010d1-10.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245844/graphs2-1010d1-10.txt)
[graphs2-1010d1-11.txt](https://github.com/openmicroscopy/openmicroscopy/files/1245841/graphs2-1010d1-11.txt)

mtbc commented:
Indeed! I am heartened by the "... DatasetAnnotationLink ... links to ..." lines because those note the results of database queries.

manics commented:
I just took a look at the postgres queries (`tail -n0 -F /var/lib/pgsql/9.4/data/pg_log/postgresql-Wed.log | grep annotationlink --line-buffered`)
```
$ wc pg-1010d1-*
     174    5253   52126 pg-1010d1-00.log
      30     969    9820 pg-1010d1-01.log
     174    5253   52126 pg-1010d1-10.log
      30     969    9820 pg-1010d1-11.log
```
`for f in pg-1010d1-*log; do echo; echo $f; tr -cs '[:alnum:]' '\n' < $f | grep annotationlink | sort | uniq -c; done`:
```

pg-1010d1-00.log
   9 annotationannotationlink
   3 channelannotationlink
  60 datasetannotationlink
   3 detectorannotationlink
   3 dichroicannotationlink
   3 experimenterannotationlink
   3 experimentergroupannotationlink
   3 filesetannotationlink
   3 filterannotationlink
   3 folderannotationlink
   3 imageannotationlink
   3 instrumentannotationlink
   3 lightpathannotationlink
   3 lightsourceannotationlink
   3 namespaceannotationlink
   3 nodeannotationlink
   3 objectiveannotationlink
   3 originalfileannotationlink
   3 planeinfoannotationlink
   3 plateacquisitionannotationlink
   3 plateannotationlink
  30 projectannotationlink
   3 reagentannotationlink
   3 roiannotationlink
   3 screenannotationlink
   3 sessionannotationlink
   3 shapeannotationlink
   3 wellannotationlink

pg-1010d1-01.log
  30 datasetannotationlink

pg-1010d1-10.log
   9 annotationannotationlink
   3 channelannotationlink
  60 datasetannotationlink
   3 detectorannotationlink
   3 dichroicannotationlink
   3 experimenterannotationlink
   3 experimentergroupannotationlink
   3 filesetannotationlink
   3 filterannotationlink
   3 folderannotationlink
   3 imageannotationlink
   3 instrumentannotationlink
   3 lightpathannotationlink
   3 lightsourceannotationlink
   3 namespaceannotationlink
   3 nodeannotationlink
   3 objectiveannotationlink
   3 originalfileannotationlink
   3 planeinfoannotationlink
   3 plateacquisitionannotationlink
   3 plateannotationlink
  30 projectannotationlink
   3 reagentannotationlink
   3 roiannotationlink
   3 screenannotationlink
   3 sessionannotationlink
   3 shapeannotationlink
   3 wellannotationlink

pg-1010d1-11.log
  30 datasetannotationlink
```

manics commented:
The difference in performance is much more noticeable when the number of parents is increased.
5000 parents, 10 annotations, delete one annotation link:
- no options: 44.04 s
- childOptions excludeType: 2.38 s
- typesToIgnore: 0.12 s

```
$ python /tmp/test_delete_performance-d1.py 5000 10 0 0
Setup complete
Continue?
Deleting 1 DatasetAnnotationLinks ([] [])
Deleted: ome.model.annotations.DatasetAnnotationLink: 1
in 44.035630 s
5000,10,44.035629987716675,['ome.model.annotations.DatasetAnnotationLink: 1']

$ python /tmp/test_delete_performance-d1.py 5000 10 1 0
Setup complete
Continue?
Deleting 1 DatasetAnnotationLinks ([object#0(::omero::cmd::graphs::ChildOption){includeType={}excludeType={[0]=Annotation}includeNs={}excludeNs={}}] [])
Deleted: ome.model.annotations.DatasetAnnotationLink: 1
in 2.376793 s
5000,10,2.3767929077148438,['ome.model.annotations.DatasetAnnotationLink: 1']

$ python /tmp/test_delete_performance-d1.py 5000 10 0 1
Setup complete
Continue?
Deleting 1 DatasetAnnotationLinks ([] ['Annotation'])
Deleted: ome.model.annotations.DatasetAnnotationLink: 1
in 0.124176 s
5000,10,0.1241757869720459,['ome.model.annotations.DatasetAnnotationLink: 1']
```
Tests were repeated with similar results.

Good to merge.

mtbc commented:
Thank you very much for all the testing.

joshmoore commented:
Are any of these tests candidates for inclusion with this PR? It seems appropriate to have something which uses the new feature actually in the code base.

mtbc commented:
Not easily as they test performance. Some kind of performance test might be plausible, though, not sure how slow it would have to be to reliably pass CI. Normally when they turn out to be flaky we just disable them.

I could write a couple of new tests that it's having some kind of `excludeType` effect.

joshmoore commented:
Tests of the accuracy would certainly be appreciated. Alternatively, we introduce this along with @manics parallel PR but on the metadata53 branch until we know more.

mtbc committed:
add integration tests for Delete2.typesToIgnore

mtbc committed:
Merge remote-tracking branch 'remotes/origin/develop' into delete-ignore-types

mtbc committed:
fix group ID argument in integration test

mtbc commented:
(The performance test takes under a couple of minutes; most of that is in setup.)

mtbc committed:
extend timeout for performance test

avoid flakiness caused by variation in CI

joshmoore commented:
Thanks for the tests. 1-2min for the [perf test](https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-java/673/testngreports/integration.delete/AnnotationDeleteTest/testIgnoreTypesPerformance/) seems reasonable, though too many of those may eventually become an issue. I have a slight concern with the changes needed in, e.g. MATLAB, to add the parameter if we're going to just remove it in the future. Other than use of Ice's `Map<String, String> context`, I'm not sure if that can be prevented though. (Using Ice's `optional` flag *might* do it, but that's untested and has protocol implications.)

mtbc commented:
In the longer term we should make such changes fine in MATLAB too: see https://trello.com/c/PvQ5yRvo/240-graph-requests-for-matlab. Happy to push a commit trying out `optional` though.

mtbc commented:
With Ice 3.6.3 the `optional` flag isn't working for me: I use `Delete2.setTypesToIgnore` on the client side but on the server in `Delete2I.init` the values are not there.

mtbc commented:
@joshmoore: Okay to merge as-is? I am guessing that the MATLAB issue is a rather minor breaking change; it's unfortunate that `optional` didn't offer an easy fix.

joshmoore commented:
Discussed today:

 * impact of API is acceptable
 * change is likely to stay in for the lifetime of 5.4
 * `DeleteData.m` might should be changed to use `deleteObjects`

dominikl commented:
Looks like you can use the `Requests` util in Matlab, too. Tested that https://github.com/mtbc/openmicroscopy/pull/15 , seems to work.

joshmoore commented:
@mtbc : do you intend to merge that PR?

mtbc commented:
Not in its current form, given @sbesson's reasonable conservatism in https://github.com/mtbc/openmicroscopy/pull/15#issuecomment-327698752, much appreciated though @dominikl's research is.

dominikl commented:
I close https://github.com/mtbc/openmicroscopy/pull/15 . Can open a separate PR. Sorry, missed the point, that the Delete example should use the deleteObjects methods. Will adjust this method then to use the Requests util methods.

joshmoore commented:
Ok. Thanks for the clarification. Merging based on https://github.com/openmicroscopy/openmicroscopy/pull/5460#issuecomment-327122506
